### PR TITLE
Update suggest feature

### PR DESCRIPTION
**Pull request by kwekke**
* Add a Logic Gate aspect to suggest possible command feature
* Add Predicate classes 
* Add tests

### CODE REVIEW

File: src/test/java/seedu/exercise/logic/commands/SuggestPossibleCommandTest.java [hunk 1]
```diff
@@ -17,15 +29,22 @@
...
 
     private Model model;
     private Model expectedModel;
     private Set<Muscle> targetMuscles;
     private Map<String, String> targetCustomProperties;
+    private Predicate<Exercise> predicateMuscleAnd;
+    private Predicate<Exercise> predicateMuscleOr;
+    private Predicate<Exercise> predicateCustomPropertyAnd;
+    private Predicate<Exercise> predicateCustomPropertyOr;
+    private boolean isStrict = true;
```
Comment (t-cheepeng) [inline src/test/java/seedu/exercise/logic/commands/SuggestPossibleCommandTest.java]: can these be local variables instead? does not seem like they need to be class variables for tests

Comment (kwekke) [reply src/test/java/seedu/exercise/logic/commands/SuggestPossibleCommandTest.java]: ok

File: src/main/java/seedu/exercise/logic/parser/SuggestCommandParser.java [hunk 2]
```diff
@@ -60,26 +65,25 @@ public SuggestCommand parse(String args) throws ParseException {
...
+
         Set<Muscle> muscles = ParserUtil.parseMuscles(argMultimap.getAllValues(PREFIX_MUSCLE));
         Map<String, String> customPropertiesMap =
                 ParserUtil.parseCustomProperties(argMultimap.getAllCustomProperties());
-        return new SuggestPossibleCommand(muscles, customPropertiesMap);
+        if (muscles.isEmpty() && customPropertiesMap.isEmpty()) {
+            throw new ParseException(String.format(MESSAGE_INVALID_COMMAND_FORMAT,
+                    SuggestCommand.MESSAGE_USAGE));
+        }
+        Predicate<Exercise> predicate = PredicateUtil.parsePredicate(muscles, customPropertiesMap, operationType);
```
[the remark this answers is not in the document]
Comment (kwekke) [reply src/main/java/seedu/exercise/logic/parser/SuggestCommandParser.java]: ok

File: src/main/java/seedu/exercise/logic/parser/predicate/ExerciseMusclePredicate.java [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+import java.util.Set;
+import java.util.function.Predicate;
+
+import seedu.exercise.model.property.Muscle;
+import seedu.exercise.model.resource.Exercise;
+
+/**
+ * Tests whether an {@code Exercise} matches the {@code predicate}
+ */
+public class ExerciseMusclePredicate implements Predicate<Exercise> {
```
[the remark this answers is not in the document]
Comment (kwekke) [reply src/main/java/seedu/exercise/logic/parser/predicate/ExerciseMusclePredicate.java]: ok

File: src/main/java/seedu/exercise/logic/parser/SuggestCommandParser.java [hunk 4]
```diff
@@ -69,7 +66,8 @@ public SuggestCommand parse(String args) throws ParseException {
      * Parses arguments and returns SuggestPossibleCommand for execution
      */
     private static SuggestCommand parsePossible(ArgumentMultimap argMultimap) throws ParseException {
-        if (!argMultimap.arePrefixesPresent(PREFIX_OPERATION_TYPE) || !argMultimap.getPreamble().isEmpty()) {
+        assert(argMultimap.getPreamble().isEmpty());
```
Comment (t-cheepeng) [inline src/main/java/seedu/exercise/logic/parser/SuggestCommandParser.java]: Why the sudden change to assertion here? This check is for user input error

Comment (kwekke) [reply src/main/java/seedu/exercise/logic/parser/SuggestCommandParser.java]: because preamble was already checked before hand in parse

Comment (t-cheepeng) [reply src/main/java/seedu/exercise/logic/parser/SuggestCommandParser.java]: Okay.

**Pull request merged**